storybooks/storybook
# Preact: Utilize `@previewjs/type-analyzer` for argType generation

Issue: #19739

## What I did

Implemented a type analyzer for Preact stories and components using `@previewjs/type-analyzer`. The type information is injected into a story file by a plugin that processes any story file (by default `/\.(story|stories)\.[tj]sx?$/`) and appends `[StoryName].argTypes = { ... }` information when types are detected for the given component/story.

The plugin itself runs in a separate worker thread to prevent the main thread from being blocked when the types are analysed. Since this only runs for files as stories, the response should be rather quick (provided the dependency graph for the story isn't too deep and wide).

The previous custom implementation was able to extract JSDoc comments, but `@previewjs/type-analyzer` doesn't support this yet, hence why the `description` field is not populated when mapping types in the plugin. This will be a future enhancement added to the `@previewjs` libraries, which this plugin can then take advantage of in a future update.

Note that `argTypes` can be overridden by the user settings `argTypes` for their story, since the original `argTypes` (if any) are spread over the generated ones to ensure the controls are overridable.

## How to test

1. Build a sandbox using the `preact-vite/default-ts` template:
   `yarn task --task sandbox --no-link --template preact-vite/default-ts --start-from=compile`
2. Run the generated sandbox.
3. In the sidebar, navigate to the `stories/components/Typed` example and note the controls being populated despite the source file not defining `argTypes` or `args`.
4. Open dev tools and navigate to **Sources** to find the `iframe` sources under `src/stories/components/Typed.stories.tsx` and check that `argTypes` are being injected for the story under the comment `;// generated argTypes from plugin`.

Example output:

<img width="367" alt="image" src="https://user-images.githubusercontent.com/112161/226603224-274c7c6d-5d2a-49ce-8ae0-6d9ea59b36c8.png">

Rendered TSX story:

<img width="1140" alt="image" src="https://user-images.githubusercontent.com/112161/226603352-dc71867f-e001-4c9c-880e-92f3425f1e4d.png">


      _Not sure yet._
      _There is a new example in the `preact` template that demonstrates type inferrence._
      _Yes. Types will now be inferred by default without requiring an external package for Preact._
